internal/nomad/nomad.go

```diff
@@ -510,3 +528,9 @@ func setUserCommand(command []string, privilegedExecution bool) []string {
 func stderrFifo(id int64) string {
 	return fmt.Sprintf(stderrFifoFormat, id)
 }
+
+func setInnerDebugMessages(command string) (result string) {
+	result = fmt.Sprintf(timeDebugMessageFormatStart, "innerCommand", command)
+	result = strings.TrimSuffix(result, ";")
+	return fmt.Sprintf(timeDebugMessageFormatEnd, result, "End")
```

Comment:
Similarly, forwarding the list of files with the nullIO writer (and the end) is still cumbersome: 
![Bildschirm­foto 2023-02-17 um 19 03 35](https://user-images.githubusercontent.com/7300329/219741582-248b8554-8268-471f-bab9-8a545f481c23.png)

Reply:
> Sure, it's needed to identify when the "innerCommand" finished, right?

Yes, this is the main purpose.

> Is the "end" command also used to determine the exit status of a program?

If the exit status of the program is not zero, the end message will not be send. This is due to that we do not want to manipulate the overall exit code returned by Nomad. If we would connect the end-echo-command with an `;` Nomad would always return 0 as exit code. Therefore, we connect those commands with `&&`. Probably, we could design more advanced/complex bash commands that would always send the exit debug message, contain the exit code in it and preserve the exit code for Nomad.

> And why could it happen that the end is as long as the "innerCommand" and not shown in Sentry the last?

It was unintuitive that the end is as long as the "innerCommand". I changed it, thanks!

Why the `nomad.execute.pipe - End` can be shown before the `nomad.execute.bash - innerCommand` is due to "the difference between `nomad.execute.pipe` and `nomad.execute.bash`".

"The one is for piping the command to the tty and the other for invoking it" - No, we pipe the command as one whole to the tty.

When the `nomad.execute.pipe`-span starts, the command is invoked. When the `nomad.execute.pipe`-span ends, the debug message is received by Poseidon. So, the duration of the `nomad.execute.pipe`-span should be only the network(aka Nomad Forwarding)-delay.
The `nomad.execute.bash`-span starts directly after the end of the `nomad.execute.pipe`-span and ends when the next `nomad.execute.pipe`-span is started (Behavior after the next commit). Therefore, the duration of the sum of the belonging `nomad.execute.pipe` and `nomad.execute.bash`-spans is the duration of the command runtime.

The end-echo message can be invoked before Poseidon receives the innerCommand-echo message. Therefore and because Sentry sort the spans after the Start Time (and not the End Time), the pipe.end span can be before the bash.innerCommand span.

> Similarly, forwarding the list of files with the nullIO writer

I explain it myself that the bash execution has finished, but the data (listing of the fs) is still forwarded/streamed through the network chain (nomad agent, nomad server, Poseidon, CodeOcean).

Reply:
> `bash -c "echo 'Test' && exit $?"`

The `?` variable is connected to the last executed command, but not to the last part of the same command separated by a `;`. Here a small example:

``` bash
$ true
$ echo $?
0
$ false
$ echo $?
1
$ true
$ bash -c "false; echo $?"
0
$ true
$ bash -c "false && echo $?"
$ true
$ bash -c "false; echo Debug; echo $?"
Debug
0
```
A solution might be constructed with an if condition:
``` bash
$ bash -c "if false; then echo Debug; true; else echo Debug; false; fi"
Debug
$ echo $?
1
```

> However, then I am wondering, why we see the `nomad.execute.pipe` multiple times?

We see the pipe-span for each debug statement (currently `SetUserAndUnsetEnv`, `innerCommand` and `End`) because it measures the time from the debug statement execution to Poseidon receiving the debug statement output.

> I am really wondering why we are missing double quotes after the bash -c?

This is due to the casting/stringify of the logger. We do pass an array as the command (exec form). Therefore, everything after the [`bash`, `-c`] is passed as a separate parameter (to libc). The separation of the individual items of the array are not visible with this logging :|

> Hm, that's surprising. Why is that the case, or is it connected to the pipelining I don't get yet?

Hm, maybe we can discuss tomorrow the common understanding of the `nomad.execute.pipe`-span and how we can improve the documentation.

Reply:
> The `?` variable is connected to the last executed command, but not to the last part of the same command separated by a `;`. 

That's exactly what I wanted. What if we have the following structure (simplified, not tested):

```shell
bash -c "
            /sbin/setuser user make run;
            bash -c \"echo 'Debug'; exit $?\"
        "
```

Wouldn't this always execute the `make` command and then always print "Debug" while still reporting the same exit code? So, I would suggest replacing the ` && echo "End"` by `; bash -c "echo 'End'; exit $?"` (wrapping this specific part in yet another bash).

> The separation of the individual items of the array are not visible with this logging :|

Ah, I see. I already assume something like that, but would (generally speaking) still like to have some debug functionality for the overall command sent to Nomad.

> Hm, maybe we can discuss tomorrow the common understanding of the `nomad.execute.pipe`-span and how we can improve the documentation.

Yes, let's do that. I think, it might get clearer already but still, I feel we should continue discussing that.

Reply:
As tested today:

```shell
bash -c "make run; bash -c \"echo 'End' && exit \$?\""
```

Reply:
The `echo 'End'` part is simplified as the `End`-message contains the `$(date +%%s%%3N)` execution. This changes the `$?` value. Also, we cannot use the simple quotes here because we already use them in an outer layer. It seems like it works by introducing a variable:

https://github.com/openHPI/poseidon/blob/d93c4afdf32ecc55e481b20aaccc7291356042dc/internal/nomad/sentry_debug_writer.go#L18

Reply:
Thanks for having a look; that sounds great and is an elegant solution, I think! 👏